pkg/arvo/app/chat-cli.hoon

```diff
@@ -0,0 +1,1111 @@
+::  chat-cli: cli chat client using chat-store and friends
+::
+::    pulls all known messages into a single stream.
+::    type ;help for usage instructions.
+::
+::    note that while the chat-store only cares about paths,
+::    we mostly deal with [ship path] (aka target) here.
+::    when sending messages (through the chat hook),
+::    we concat the ship onto the head of the path,
+::    and trust it to take care of the rest.
+::
+/-  sole-sur=sole, *chat-store, *chat-view, *chat-hook
+/+  sole-lib=sole
+/=  seed  /~  !>(.)
+::
+|%
++$  state
+  $:  grams=(list mail)                             ::  all messages
```

Comment:
Why do you need to duplicate state here? It seems messy. Why not just scry?

Reply:
talk/chat-cli aggregates messages into a single stream. The user interacts with the messages therein as if they're all part of a single stream. This does not map cleanly onto the chat-store, which (I believe) doesn't let me say "give me the third to last message" across everything at once.

Maybe chat-cli needs to try to present a more "per-path style" interface. It probably shouldn't start attempting that right now.

Reply:
I have a very bad feeling about the idea of allowing this code to be merged while it still duplicates state in this way.

Reply:
Then please elaborate why it gives you a very bad feeling. I have nothing more to go on/fight against than "it seems messy".

Current talk does this and hasn't exploded, even though it needs to support edits to its copy of the state, which chat-store doesn't even support.

Reply:
It's not that I think it will explode, but rather that one of the main advantages within the 'store' model that we're using is that scries can be used in the place of extraneous state, and that the %chat-cli and %chat-view are both views, or ways of rendering, the same source of truth (%chat-store). Having %chat-cli store this data adds another source of truth, and given how similar the data stores are to each other (you're not doing any intense transforms on the data) it seems unnecessary and messy.

Reply:
`chat-store` doesn't maintain this information, though, right?  Specifically, it doesn't maintain a total ordering of messages across multiple streams.  This is not intrinsic to `chat-store`, as evidenced by the fact that `chat-view` doesn't use it (since it doesn't merge them all into a single stream).  Since `chat-cli` uses a total order of messages for things like referring to messages by number, it's not unreasonable for it to somewhere maintain that state.  Note this information is local and non-permanent in the sense that eg you could throw it away on upgrade if you resubscribed as well.

Reply:
Yes, you're right, Phil. It seems that I'm arguing mostly against the idea that %chat-cli should keep a total ordering of messages given that it requires this state duplication.